static/js/components/email/hoc.js

```diff
@@ -0,0 +1,120 @@
+      }
+
+      getActiveEmailState(): EmailState {
+        const { email } = this.props;
+        return email[email.currentlyActive];
+      }
+
+      openEmailComposer = R.curry((emailType: string, emailOpenParams: any): void => {
+        const { dispatch } = this.props;
+        dispatch(
+          startEmailEdit({
+            type: emailType,
+            ...emailConfigs[emailType].emailOpenAction(emailOpenParams)
+          })
+        );
+        dispatch(showDialog(EMAIL_COMPOSITION_DIALOG));
+      });
+
+      updateEmailFieldEdit = R.curry((fieldName, e): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        let activeEmail = this.getActiveEmailState();
+        let inputsClone = R.clone(activeEmail.inputs);
+        inputsClone[fieldName] = e.target.value;
+        dispatch(updateEmailEdit({type: currentlyActive, inputs: inputsClone}));
+        if (!R.isEmpty(activeEmail.validationErrors)) {
+          let cloneErrors = emailValidation(inputsClone);
+          dispatch(updateEmailValidation({type: currentlyActive, errors: cloneErrors}));
+        }
+      });
+
+      closeAndClearEmailComposer = (): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        dispatch(clearEmailEdit(currentlyActive));
+        dispatch(hideDialog(EMAIL_COMPOSITION_DIALOG));
+      };
+
+      closeEmailComposerAndSend = (): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        let activeEmail = this.getActiveEmailState();
+        let errors = emailValidation(activeEmail.inputs);
+        dispatch(updateEmailValidation({type: currentlyActive, errors: errors}));
+        if (R.isEmpty(errors)) {
+          dispatch(
+            emailConfigs[currentlyActive].emailSendAction(activeEmail)
+          ).then(() => {
+            this.closeAndClearEmailComposer();
+          });
+        }
+      };
+
+      renderCompositionDialog(): ?React$Element<*> {
+        let { ui: { dialogVisibility } } = this.props;
+
+        let activeEmailType = this.getActiveEmailType();
+
+        if (!activeEmailType) return null;
+
+        return (
+          <EmailCompositionDialog
+            updateEmailFieldEdit={this.updateEmailFieldEdit}
+            closeAndClearEmailComposer={this.closeAndClearEmailComposer}
+            closeEmailComposerAndSend={this.closeEmailComposerAndSend}
+            dialogVisibility={dialogVisibility[EMAIL_COMPOSITION_DIALOG]}
+            activeEmail={this.getActiveEmailState()}
+            title={emailConfigs[activeEmailType].title}
+            subheadingRenderer={emailConfigs[activeEmailType].renderSubheading}
+          />
+        );
+      }
+
+      render() {
+        let emailOpenActions = {};
+        for (let emailType of Object.keys(emailConfigs)) {
+          emailOpenActions[emailType] = this.openEmailComposer(emailType);
+        }
```

Comment:
so take [this line](https://github.com/mitodl/micromasters/pull/2555/files#diff-19b75ce18fd1eaefee48fc255d15b45bR153) from `LearnerSearch.js` as an example:

```javascript
onClick={R.partial(openSearchResultEmailComposer, [this.searchkit])}
// would become...
onClick={R.partial(openSearchResultEmailComposer, [SEARCH_EMAIL_TYPE, this.searchkit])}
```

is that what you had in mind? i'm not sure about this. i understand the drawback of the current approach - we are partially satisfying the arguments of the function two times, which can make it confusing to figure out what the function needs at any given point during the run time. this change would introduce some problems too. the email type value on the container page would need to match the value at the call site, and there's no way we could 'enforce' those two values being the same. i would rather have the slightly more confusing invocation pattern than open the potential for a bug where the email types don't match. the former can be somewhat easily figured out with a debugger or console.log lines, but the latter could send someone on a very lengthy search across several files to figure out why their expected dialog isn't coming up. we know what type of email we want to open when this function is passed down to the wrapped component's props from the HOC, so i think it makes enough sense to satisfy that argument up front.